### PR TITLE
ENH/BUG: Use Kleene logic for groupby any/all

### LINKED ISSUE #33449: BUG: Series.any/all with level keyword should preserve boolean dtype
```
In [5]: s = pd.Series([False, False, True, True, False, True],
                      index=[0, 0, 1, 1, 2, 2], dtype="boolean")  

In [6]: s.all(level=0) 
Out[6]: 
0    False
1     True
2    False
dtype: bool
```

The dtype of the result could also be boolean instead of bool.

### LINKED ISSUE #37506: ENH: implement Kleene versions of GroupBy any/all kernels for nullable dtypes
xref   https://github.com/pandas-dev/pandas/issues/37493#issuecomment-718974981

### LINKED ISSUE #40585: BUG: Groupby.any(skipna=True)
- [x] I have checked that this issue has not already been reported.

- [x] I have confirmed this bug exists on the latest version of pandas.

- [ ] (optional) I have confirmed this bug exists on the master branch of pandas.

---

**Note**: Please read [this guide](https://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports) detailing how to provide the necessary information for us to reproduce your bug.

#### Code Sample, a copy-pastable example

```python
import pandas as pd


def test_groupby_any():
    df = pd.DataFrame(
        {
            "ids": ["id1", "id1", "id2", "id2"],
            "A": [True, False, pd.NA, True]
        }
    ).astype({"ids": pd.StringDtype(),
              "A": pd.BooleanDtype()})

    result = df.groupby("ids").any(skipna=True)

    expected = pd.DataFrame(
        {
            "A": [True, True]
        }, index=pd.Series(["id1", "id2"], name="ids")
    )

    pd.testing.assert_frame_equal(expected, result)

```

#### Problem description

When running df.groupby(field_to_group_on).any(skipna=True) it should be expected that pd.NA values are neglected for the .any() method. However the output is this:

self = <BooleanArray>
[True, False, <NA>, True]
Length: 4, dtype: boolean
dtype = dtype('bool'), copy = True

    def astype(self, dtype, copy: bool = True) -> ArrayLike:
        """
        Cast to a NumPy array or ExtensionArray with 'dtype'.
    
        Parameters
        ----------
        dtype : str or dtype
            Typecode or data-type to which the array is cast.
        copy : bool, default True
            Whether to copy the data, even if not necessary. If False,
            a copy is made only if the old dtype does not match the
            new dtype.
    
        Returns
        -------
        ndarray or ExtensionArray
            NumPy ndarray, BooleanArray or IntegerArray with 'dtype' for its dtype.
    
        Raises
        ------
        TypeError
            if incompatible type with an BooleanDtype, equivalent of same_kind
            casting
        """
        from pandas.core.arrays.string_ import StringDtype
    
        dtype = pandas_dtype(dtype)
    
        if isinstance(dtype, BooleanDtype):
            values, mask = coerce_to_array(self, copy=copy)
            if not copy:
                return self
            else:
                return BooleanArray(values, mask, copy=False)
        elif isinstance(dtype, StringDtype):
            return dtype.construct_array_type()._from_sequence(self, copy=False)
    
        if is_bool_dtype(dtype):
            # astype_nansafe converts np.nan to True
            if self._hasna:
>               raise ValueError("cannot convert float NaN to bool")
E               ValueError: cannot convert float NaN to bool

#### Expected Output

```
    expected = pd.DataFrame(
        {
            "A": [True, True]
        }, index=pd.Series(["id1", "id2"], name="ids")
    )
```
#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit           : f2c8480af2f25efdbd803218b9d87980f416563e
python           : 3.9.1.final.0
python-bits      : 64
OS               : Windows
OS-release       : 10
Version          : 10.0.19041
machine          : AMD64
processor        : Intel64 Family 6 Model 85 Stepping 4, GenuineIntel
byteorder        : little
LC_ALL           : None
LANG             : None
LOCALE           : English_United States.1252
pandas           : 1.2.3
numpy            : 1.20.1
pytz             : 2021.1
dateutil         : 2.8.1
pip              : 21.0.1
setuptools       : 52.0.0
Cython           : None
pytest           : 6.2.2
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : None
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : 2.11.3
IPython          : None
pandas_datareader: None
bs4              : None
bottleneck       : None
fsspec           : None
fastparquet      : None
gcsfs            : None
matplotlib       : None
numexpr          : None
odfpy            : None
openpyxl         : 3.0.7
pandas_gbq       : None
pyarrow          : None
pyxlsb           : None
s3fs             : None
scipy            : 1.6.1
sqlalchemy       : 1.4.2
tables           : None
tabulate         : None
xarray           : None
xlrd             : 1.2.0
xlwt             : None
numba            : None

</details>

### PR DESCRIPTION
**Pull request by mzeitlin11**
- [x] closes #37506
- [x] closes #40585
- [x] closes #33449 
- [x] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them
- [x] whatsnew entry

Built on #40811

ASV's don't look affected:

<details>

```
before           after         ratio
     [3b4b193d]       [423f43fc]
     <master>         <enh/any_all_kleene>
          216±8ms          205±4ms     0.95  groupby.GroupByCythonAgg.time_frame_agg('float64', 'all')
          208±8ms          197±3ms     0.95  groupby.GroupByCythonAgg.time_frame_agg('float64', 'any')
         162±10μs          139±4μs    ~0.86  groupby.GroupByMethods.time_dtype_as_field('datetime', 'all', 'direct')
         161±20μs          138±9μs    ~0.86  groupby.GroupByMethods.time_dtype_as_field('datetime', 'all', 'transformation')
         159±10μs          142±5μs    ~0.90  groupby.GroupByMethods.time_dtype_as_field('datetime', 'any', 'direct')
         160±10μs          140±5μs    ~0.88  groupby.GroupByMethods.time_dtype_as_field('datetime', 'any', 'transformation')
          109±8μs         113±10μs     1.04  groupby.GroupByMethods.time_dtype_as_field('float', 'all', 'direct')
         103±10μs          109±9μs     1.06  groupby.GroupByMethods.time_dtype_as_field('float', 'all', 'transformation')
         113±20μs         132±10μs    ~1.17  groupby.GroupByMethods.time_dtype_as_field('float', 'any', 'direct')
         138±10μs         124±20μs    ~0.90  groupby.GroupByMethods.time_dtype_as_field('float', 'any', 'transformation')
          110±9μs          112±5μs     1.02  groupby.GroupByMethods.time_dtype_as_field('int', 'all', 'direct')
          115±9μs         109±10μs     0.94  groupby.GroupByMethods.time_dtype_as_field('int', 'all', 'transformation')
         123±10μs          115±7μs     0.94  groupby.GroupByMethods.time_dtype_as_field('int', 'any', 'direct')
         125±10μs          114±8μs     0.92  groupby.GroupByMethods.time_dtype_as_field('int', 'any', 'transformation')
         449±30μs         452±20μs     1.01  groupby.GroupByMethods.time_dtype_as_field('object', 'all', 'direct')
         442±30μs         445±20μs     1.01  groupby.GroupByMethods.time_dtype_as_field('object', 'all', 'transformation')
         470±10μs         455±30μs     0.97  groupby.GroupByMethods.time_dtype_as_field('object', 'any', 'direct')
         466±30μs          406±4μs    ~0.87  groupby.GroupByMethods.time_dtype_as_field('object', 'any', 'transformation')
          118±8μs          118±7μs     1.00  groupby.GroupByMethods.time_dtype_as_group('datetime', 'all', 'direct')
          134±9μs         132±20μs     0.99  groupby.GroupByMethods.time_dtype_as_group('datetime', 'all', 'transformation')
         131±20μs          115±9μs    ~0.88  groupby.GroupByMethods.time_dtype_as_group('datetime', 'any', 'direct')
         121±10μs          118±5μs     0.97  groupby.GroupByMethods.time_dtype_as_group('datetime', 'any', 'transformation')
         111±10μs          111±4μs     1.00  groupby.GroupByMethods.time_dtype_as_group('float', 'all', 'direct')
         109±10μs         120±10μs    ~1.11  groupby.GroupByMethods.time_dtype_as_group('float', 'all', 'transformation')
          110±9μs         127±10μs    ~1.16  groupby.GroupByMethods.time_dtype_as_group('float', 'any', 'direct')
         112±10μs         122±10μs     1.09  groupby.GroupByMethods.time_dtype_as_group('float', 'any', 'transformation')
         130±10μs          115±6μs    ~0.89  groupby.GroupByMethods.time_dtype_as_group('int', 'all', 'direct')
          133±7μs          115±7μs    ~0.86  groupby.GroupByMethods.time_dtype_as_group('int', 'all', 'transformation')
          117±5μs         123±10μs     1.05  groupby.GroupByMethods.time_dtype_as_group('int', 'any', 'direct')
         133±10μs         118±20μs    ~0.89  groupby.GroupByMethods.time_dtype_as_group('int', 'any', 'transformation')
          107±9μs         122±10μs    ~1.14  groupby.GroupByMethods.time_dtype_as_group('object', 'all', 'direct')
         113±10μs          107±5μs     0.94  groupby.GroupByMethods.time_dtype_as_group('object', 'all', 'transformation')
         135±10μs          108±4μs    ~0.80  groupby.GroupByMethods.time_dtype_as_group('object', 'any', 'direct')
          114±6μs          105±2μs     0.92  groupby.GroupByMethods.time_dtype_as_group('object', 'any', 'transformation')
```

</details>

### CODE REVIEW

File: pandas/tests/reductions/test_reductions.py [hunk 1]
```diff
@@ -949,14 +949,17 @@ def test_all_any_boolean(self):
...
-        # Check level TODO(GH-33449) result should also be boolean
         s = Series(
             [False, False, True, True, False, True],
             index=[0, 0, 1, 1, 2, 2],
             dtype="boolean",
         )
-        tm.assert_series_equal(s.all(level=0), Series([False, True, False]))
-        tm.assert_series_equal(s.any(level=0), Series([False, True, True]))
+        tm.assert_series_equal(
+            s.all(level=0), Series([False, True, False], dtype="boolean")
```
Comment (rhshadrach) [inline pandas/tests/reductions/test_reductions.py:958]: Can you set result/expected on separate lines? This way it's easier to step through with a debugger.

File: pandas/core/groupby/groupby.py [hunk 2]
```diff
@@ -2735,6 +2746,10 @@ def _get_cythonized_result(
                 mask = isna(values).view(np.uint8)
                 func = partial(func, mask)
 
+            if how == "group_any_all" and isinstance(values, BaseMaskedArray):
+                post_processing = partial(post_processing, result_is_nullable=True)
+                func = partial(func, use_kleene_logic=True)
```
Comment (rhshadrach) [inline pandas/core/groupby/groupby.py:2751]: It seems preferable to me to not special case here. Can use partial on post_processing in the caller. For func, would we instead use set `any, all, any_kleene, all_kleene` for val_test instead of an additional parameter?

**Review by rhshadrach — CHANGES_REQUESTED** [review]
Thanks for the PR!

File: doc/source/whatsnew/v1.3.0.rst [hunk 3]
```diff
@@ -195,6 +195,8 @@ Other enhancements
 - :class:`RangeIndex` can now be constructed by passing a ``range`` object directly e.g. ``pd.RangeIndex(range(3))`` (:issue:`12067`)
 - :meth:`round` being enabled for the nullable integer and floating dtypes (:issue:`38844`)
 - :meth:`pandas.read_csv` and :meth:`pandas.read_json` expose the argument ``encoding_errors`` to control how encoding errors are handled (:issue:`39450`)
+- :class:`SeriesGroupBy` and :class:`DataFrameGroupBy` return a ``BooleanDType`` result using Kleene logic for methods ``any`` and ``all`` with nullable data types (:issue:`37506`, :issue:`40585`, :issue:`33449`)
```
[the remark this answers is not in the document]
Comment (mzeitlin11) [reply doc/source/whatsnew/v1.3.0.rst]: Makes sense, split description up

File: pandas/tests/reductions/test_reductions.py (hunk 1 above)
Comment (mzeitlin11) [reply pandas/tests/reductions/test_reductions.py]: Yep sounds good, just ended up breaking off into a new test

File: pandas/core/groupby/groupby.py (hunk 2 above)
Comment (mzeitlin11) [reply pandas/core/groupby/groupby.py]: Thanks for the great suggestions! 

The problem I was running into is that the masked argument in `post_processing` and whether or not to use Kleene logic are not invariant in the loop in `_get_cythonized_result`. The caller doesn't know the nullable status because it might be possible that the first column is nullable, but the second column is not.

Definitely agree on trying to avoid special-casing, changed to provide a more general mechanism for nullable-dependent processing like this. While it would only be used for this case now, should make it easier to fix other nullable cases, such as std() or quantile() returning a float instead of Float.

Your suggestion for `any, all, any_kleene, all_kleene` instead of using a new parameter would be a neat way to avoid adding an additional parameter. However, couldn't figure out a clean way to do this without special casing any/all or adding more invasive mask-specific dispatch (any suggestions on this welcome!)

File: pandas/tests/groupby/test_any_all.py [hunk 4]
```diff
@@ -68,3 +69,83 @@ def test_bool_aggs_dup_column_labels(bool_agg_func):
...
+)
+def test_masked_kleene_logic(bool_agg_func, data, expected, skipna):
+    # GH#37506
+    df = DataFrame(data, dtype="boolean")
+    expected = DataFrame(
+        [expected[skipna][bool_agg_func == "all"]], dtype="boolean", index=[1]
+    )
+
+    result = df.groupby([1, 1, 1]).agg(bool_agg_func, skipna=skipna)
+    tm.assert_frame_equal(result, expected)
```
Comment (jorisvandenbossche) [inline pandas/tests/groupby/test_any_all.py:96]: An additional check could be to also calculate it on the full column, which needs to be the same here (`df[0].any(skipna=skipna)`, as an extra verification that we correctly wrote up those expected results)

**Review by jorisvandenbossche — COMMENTED** [review]
@mzeitlin11 Thanks a lot for working on this!

File: pandas/tests/groupby/test_any_all.py (hunk 4 above)
Comment (mzeitlin11) [reply pandas/tests/groupby/test_any_all.py:97]: Added this check

File: pandas/core/groupby/groupby.py [hunk 5]
```diff
@@ -1435,6 +1446,7 @@ def result_to_bool(result: np.ndarray, inference: type) -> np.ndarray:
             post_processing=result_to_bool,
             val_test=val_test,
             skipna=skipna,
+            masked=False,
```
[the remark this answers is not in the document]
Comment (mzeitlin11) [reply pandas/core/groupby/groupby.py]: Had originally used `masked` being included in kwargs as a signal for whether to pass `masked` into the groupby func. But the original value here wasn't relevant since it gets set before being used.

This is definitely confusing and a poor solution, changed to instead use a positional arg

[the remark this answers is not in the document]
Comment (jorisvandenbossche) [reply pandas/core/groupby/groupby.py]: Thanks, the update looks clearer to follow now

File: pandas/tests/groupby/test_any_all.py [hunk 6]
```diff
@@ -85,16 +86,62 @@ def test_bool_aggs_dup_column_labels(bool_agg_func):
...
+        (
+            "Float64",
+            "boolean",
+            pd.array([pd.NA], dtype="boolean"),
+            pd.array([pd.NA], dtype="boolean"),
+        ),
+    ],
+)
+def test_masked_mixed_types(dtype1, dtype2, exp_col1, exp_col2):
+    data = [1.0, np.nan]
```
Comment (rhshadrach) [inline pandas/tests/groupby/test_any_all.py:136]: Issue # here

File: pandas/core/groupby/groupby.py [hunk 7]
```diff
@@ -2674,7 +2676,11 @@ def _get_cythonized_result(
             Function to be applied to result of Cython function. Should accept
             an array of values as the first argument and type inferences as its
             second argument, i.e. the signature should be
-            (ndarray, Type).
+            (ndarray, Type). Optionally, a third argument can be "masked", to
+            allow for processing specific to nullable values
+        use_nullable_input_arg : bool, default False
+            If True, pass an argument "masked" to the cython function specifying
+            whether or not the input should be treated as masked
```
Comment (rhshadrach) [inline pandas/core/groupby/groupby.py:2683]: For consistency, with the rest of the args here, would it make sense to call this `needs_nullable`?

File: pandas/_libs/groupby.pyx [hunk 8]
```diff
@@ -412,16 +414,20 @@ def group_any_all(uint8_t[::1] out,
...
         Flag to ignore nan values during truth testing
+    masked : bool
+        If True, compute the result using Kleene logic
 
     Notes
     -----
     This method modifies the `out` parameter rather than returning an object.
-    The returned values will either be 0 or 1 (False or True, respectively).
+    The returned values will either be 0, 1 (False or True, respectively), or
+    2 to signify a masked position in the case of a nullable input.
```
Comment (jreback) [inline pandas/_libs/groupby.pyx:424]: hmm is this what we do elsewhere to encode Kleene logic? (e.g. would rather use -1 on nulls)

File: pandas/core/groupby/groupby.py (hunk 7 above)
Comment (rhshadrach) [reply pandas/core/groupby/groupby.py:2683]: And group with those args as well.

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/_libs/groupby.pyx [hunk 9]
```diff
@@ -414,8 +414,8 @@ def group_any_all(uint8_t[::1] out,
         String object dictating whether to use any or all truth testing
     skipna : bool
         Flag to ignore nan values during truth testing
-    use_kleene_logic : bool, default False
-        Whether or not to compute the result using Kleene logic
+    masked : bool
```
Comment (rhshadrach) [inline pandas/_libs/groupby.pyx:417]: I'd suggest calling this `is_nullable` instead, with the docstring explaining that we use kleene logic on nullable dtypes.

**Review by rhshadrach — CHANGES_REQUESTED** [review]
Thanks @mzeitlin11, I see now why we need logic within `_get_cythonized_result`. Really like the adjustment made there. Minor comments, otherwise lgtm.

File: pandas/tests/groupby/test_any_all.py (hunk 6 above)
Comment (mzeitlin11) [reply pandas/tests/groupby/test_any_all.py:137]: done

File: pandas/core/groupby/groupby.py (hunk 7 above)
Comment (mzeitlin11) [reply pandas/core/groupby/groupby.py]: Thanks, your naming suggestions are much clearer

File: pandas/_libs/groupby.pyx (hunk 9 above)
Comment (mzeitlin11) [reply pandas/_libs/groupby.pyx]: done

File: pandas/_libs/groupby.pyx (hunk 8 above)
Comment (mzeitlin11) [reply pandas/_libs/groupby.pyx]: The other algo is a bit different since it can be vectorized and computes both data and mask (so doesn't need to store something signifying NULL). Agree that -1 is more intuitive, have changed to use -1 as the mask signal

Comment (jorisvandenbossche) [reply pandas/_libs/groupby.pyx]: At some point, it might be interesting to look into directly filling in the mask as well instead of using the temporary -1.

Personally, I would maybe leave it at using 2, so you don't need to change the uint8 type

Comment (mzeitlin11) [reply pandas/_libs/groupby.pyx]: Yep I don't think that would be too bad, that's how I had approached it originally, but changes were slightly more invasive, so stuck with this.

For the -1 vs 2 question, I think I slightly prefer -1 because it seems more immediately apparent what it represents. Although it forces changing the dtype, still the same number of bytes, so memory layout is unaffected and constructing the view looks equivalent speed-wise:

```
In [1]: import numpy as np

In [2]: arr = np.array([True, False, True])

In [3]: %timeit arr.view('uint8')
332 ns ± 2.84 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)

In [4]: %timeit arr.view('int8')
334 ns ± 3.78 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
```
But not a strong opinion, fine with using 2 as well :)

Comment (jorisvandenbossche) [reply pandas/_libs/groupby.pyx]: If speed-wise the int8 vs uint8 doesn't matter, then indeed the -1 is fine as well

**Review by rhshadrach — APPROVED** [review]
lgtm

File: doc/source/whatsnew/v1.3.0.rst [hunk 10]
```diff
@@ -213,6 +213,9 @@ Other enhancements
 - :class:`RangeIndex` can now be constructed by passing a ``range`` object directly e.g. ``pd.RangeIndex(range(3))`` (:issue:`12067`)
 - :meth:`round` being enabled for the nullable integer and floating dtypes (:issue:`38844`)
 - :meth:`pandas.read_csv` and :meth:`pandas.read_json` expose the argument ``encoding_errors`` to control how encoding errors are handled (:issue:`39450`)
+- :class:`SeriesGroupBy` and :class:`DataFrameGroupBy` use Kleene logic for methods ``any`` and ``all`` with nullable data types (:issue:`37506`)
```
Comment (jorisvandenbossche) [inline doc/source/whatsnew/v1.3.0.rst:216]:
```suggestion
- :meth:`.GroupBy.any` and :meth:`.GroupBy.all` use Kleene logic with nullable data types (:issue:`37506`)
```

(I personally try to avoid using SeriesGroupBy and DataFrameGroupBy directly, as don't say much to many users, I think, and in this case the actual methods can be referred to)

File: doc/source/whatsnew/v1.3.0.rst [hunk 11]
```diff
@@ -213,6 +213,9 @@ Other enhancements
...
+- :class:`SeriesGroupBy` and :class:`DataFrameGroupBy` return a ``BooleanDType`` for methods ``any`` and ``all`` with nullable data types (:issue:`33449`)
```
Comment (jorisvandenbossche) [inline doc/source/whatsnew/v1.3.0.rst:217]:
```suggestion
- :meth:`.GroupBy.any` and :meth:`.GroupBy.all` return a ``BooleanDtype`` for columns with nullable data types (:issue:`33449`)
```

File: doc/source/whatsnew/v1.3.0.rst [hunk 12]
```diff
@@ -782,6 +785,8 @@ Groupby/resample/rolling
 - Bug in :meth:`Series.asfreq` and :meth:`DataFrame.asfreq` dropping rows when the index is not sorted (:issue:`39805`)
 - Bug in aggregation functions for :class:`DataFrame` not respecting ``numeric_only`` argument when ``level`` keyword was given (:issue:`40660`)
 - Bug in :class:`core.window.RollingGroupby` where ``as_index=False`` argument in ``groupby`` was ignored (:issue:`39433`)
+- Bug in :class:`SeriesGroupBy` and :class:`DataFrameGroupBy` raising ``ValueError`` when using methods ``any`` and ``all`` with nullable type columns holding ``NA`` even with ``skipna=True`` (:issue:`40585`)
```
Comment (jorisvandenbossche) [inline doc/source/whatsnew/v1.3.0.rst:788]:
```suggestion
- Bug in :meth:`.GroupBy.any` and :meth:`.GroupBy.all` raising ``ValueError`` when using with nullable type columns holding ``NA`` even with ``skipna=True`` (:issue:`40585`)
```

File: pandas/tests/groupby/test_any_all.py [hunk 13]
```diff
@@ -68,3 +69,91 @@ def test_bool_aggs_dup_column_labels(bool_agg_func):
...
+    assert (result is pd.NA and expected is pd.NA) or result == expected
+
+
+@pytest.mark.parametrize(
+    "dtype1,dtype2,exp_col1,exp_col2",
+    [
+        (
+            "float",
+            "Float64",
+            pd.array([True], dtype=bool),
```
Comment (jorisvandenbossche) [inline pandas/tests/groupby/test_any_all.py:112]:
```suggestion
            np.array([True], dtype=bool),
```

(for the resulting DataFrame, both are equivalent, but since the goal is to create a column with a numpy dtype, I find it a bit more clear / explicit to create a numpy array)

File: pandas/tests/groupby/test_any_all.py [hunk 14]
```diff
@@ -68,3 +69,91 @@ def test_bool_aggs_dup_column_labels(bool_agg_func):
...
+            pd.array([pd.NA], dtype="boolean"),
+        ),
+        (
+            "Int64",
+            "float",
+            pd.array([pd.NA], dtype="boolean"),
+            pd.array([True], dtype=bool),
```
Comment (jorisvandenbossche) [inline pandas/tests/groupby/test_any_all.py:119]:
```suggestion
            np.array([True], dtype=bool),
```

**Review by jorisvandenbossche — APPROVED** [review]
Looks good to me! (and much cleaner now ;))

I only added a few tiny stylistic/doc suggestions you can apply.

File: pandas/tests/groupby/test_any_all.py [hunk 15]
```diff
@@ -68,3 +69,91 @@ def test_bool_aggs_dup_column_labels(bool_agg_func):
...
+    expected = DataFrame(
+        [expected_data[skipna][bool_agg_func == "all"]], dtype="boolean", index=[1]
+    )
+
+    result = df.groupby([1, 1, 1]).agg(bool_agg_func, skipna=skipna)
+    tm.assert_frame_equal(result, expected)
+
+    # The expected result we compared to should match aggregating on the whole
+    # series
+    result = getattr(df[0], bool_agg_func)(skipna=skipna)
```
Comment (jreback) [inline pandas/tests/groupby/test_any_all.py:101]: this check is really hard to parse.
what exactly are you checking? can you make this much simpler

File: pandas/tests/groupby/test_any_all.py [hunk 16]
```diff
@@ -68,3 +69,91 @@ def test_bool_aggs_dup_column_labels(bool_agg_func):
...
+        ([pd.NA, pd.NA, pd.NA], [[pd.NA, pd.NA], [False, True]]),
+        ([False, pd.NA, False], [[pd.NA, False], [False, False]]),
+        ([True, pd.NA, True], [[True, pd.NA], [True, True]]),
+        ([True, pd.NA, False], [[True, False], [True, False]]),
+    ],
+)
+def test_masked_kleene_logic(bool_agg_func, data, expected_data, skipna):
+    # GH#37506
+    df = DataFrame(data, dtype="boolean")
+    expected = DataFrame(
```
Comment (jreback) [inline pandas/tests/groupby/test_any_all.py:92]: this expected is really hard to parse can you do in multiple steps / make simpler

**Review by jreback — CHANGES_REQUESTED** [review]
pls merge master as well

File: doc/source/whatsnew/v1.3.0.rst (hunk 10 above)
[the remark this answers is not in the document]
Comment (jorisvandenbossche) [reply doc/source/whatsnew/v1.3.0.rst:216]: This is in "Other enhancements", not "Other bug fixes" (so not related to https://github.com/pandas-dev/pandas/issues/40897), so I think this is the appropriate place.

File: pandas/tests/groupby/test_any_all.py (hunk 15 above)
Comment (jorisvandenbossche) [reply pandas/tests/groupby/test_any_all.py:101]: What this is essentially doing is `series.any(skipna=skipna)`, but then with `getattr` because it's generic for any/all. Do you have a concrete suggestion to change? (or to clarify the comment?)

File: doc/source/whatsnew/v1.3.0.rst (hunk 10 above)
[the remark this answers is not in the document]
Comment (jreback) [reply doc/source/whatsnew/v1.3.0.rst:216]: ahh ok

File: pandas/tests/groupby/test_any_all.py (hunk 15 above)
Comment (jreback) [reply pandas/tests/groupby/test_any_all.py:101]: there is a complicated expexcted then a really complicated assert. This needs to be greatly simplified. It is impossible to grok with all of this logic. my suggesetion is to break this up.

Comment (jorisvandenbossche) [reply pandas/tests/groupby/test_any_all.py:101]: To be very concrete, you mean breaking it up into 2 lines, like

```
series = df[0]
getattr(series, method)(skipna=skipna)
```

?

File: doc/source/whatsnew/v1.3.0.rst (hunk 10 above)
Comment (mzeitlin11) [reply doc/source/whatsnew/v1.3.0.rst:216]: Good to know, thanks

File: pandas/tests/groupby/test_any_all.py (hunk 13 above)
Comment (mzeitlin11) [reply pandas/tests/groupby/test_any_all.py:112]: Sounds good

File: pandas/tests/groupby/test_any_all.py (hunk 16 above)
Comment (mzeitlin11) [reply pandas/tests/groupby/test_any_all.py]: Addressing your comment above makes this simpler, let me know if you think any piece is still confusing

File: doc/source/whatsnew/v1.3.0.rst [hunk 17]
```diff
@@ -217,6 +217,9 @@ Other enhancements
 - :class:`RangeIndex` can now be constructed by passing a ``range`` object directly e.g. ``pd.RangeIndex(range(3))`` (:issue:`12067`)
 - :meth:`round` being enabled for the nullable integer and floating dtypes (:issue:`38844`)
 - :meth:`pandas.read_csv` and :meth:`pandas.read_json` expose the argument ``encoding_errors`` to control how encoding errors are handled (:issue:`39450`)
+- :meth:`.GroupBy.any` and :meth:`.GroupBy.all` use Kleene logic with nullable data types (:issue:`37506`)
```
Comment (jreback) [inline doc/source/whatsnew/v1.3.0.rst:220]: hmm is this how we other groupby refs? i am not sure this will render

File: pandas/tests/reductions/test_reductions.py [hunk 18]
```diff
@@ -941,31 +941,45 @@ def test_all_any_params(self):
...
+        result = getattr(ser, bool_agg_func)(skipna=skipna)
+        assert (result is pd.NA and expected is pd.NA) or result == expected
 
-        # Check level TODO(GH-33449) result should also be boolean
-        s = Series(
+    @pytest.mark.parametrize(
+        "bool_agg_func,expected",
+        [("all", [False, True, False]), ("any", [False, True, True])],
+    )
+    def test_any_all_boolean_level(self, bool_agg_func, expected):
```
Comment (jreback) [inline pandas/tests/reductions/test_reductions.py:972]: these are entirely dupilcative of the groupby ones (but i guess ok)

**Review by jreback — CHANGES_REQUESTED** [review]
ok looks good. question on the rendering in the notes, ping on green.

File: doc/source/whatsnew/v1.3.0.rst (hunk 17 above)
Comment (mzeitlin11) [reply doc/source/whatsnew/v1.3.0.rst:220]: I see one other entry with this style, I think it's good here to avoid having the duplication of `DataFrameGroupBy.`, `SeriesGroupBy.` for both `any` and `all`. Building docs now to check rendering

File: pandas/tests/reductions/test_reductions.py (hunk 18 above)
Comment (mzeitlin11) [reply pandas/tests/reductions/test_reductions.py:972]: Yeah can remove if you'd prefer, this was just to explicitly address #33449

File: doc/source/whatsnew/v1.3.0.rst (hunk 17 above)
Comment (mzeitlin11) [reply doc/source/whatsnew/v1.3.0.rst:220]: Looks good:
Screen Shot 2021-04-13 at 1 08 05 PM
(links to https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.core.groupby.GroupBy.all.html)

**Review by jreback — APPROVED** [review]

**Comment by mzeitlin11** [comment]
Greenish, failures unrelated and fixed by #40930 and flaky zip fix that just got merged

**Comment by jreback** [comment]
thanks @mzeitlin11 very nice!

File: doc/source/whatsnew/v1.3.0.rst (hunk 17 above)
Comment (jorisvandenbossche) [reply doc/source/whatsnew/v1.3.0.rst:220]: > hmm is this how we other groupby refs? i am not sure this will render

Yes, that's something @rhshadrach recently found out. The leading `.` is basically a wildcard search for a matching object, so `.GroupBy` will find `pandas.core.groupby.GroupBy` (since we only have a single `GroupBy` class in our API docs)

Comment (rhshadrach) [reply doc/source/whatsnew/v1.3.0.rst:220]: I do wonder if this is something in the docs we could find a better solution to - knowing whether to look under GroupBy or DataFrameGroupBy etc. I personally don't find it hurts usability of the docs, and is only a slight annoyance when crafting the whatsnew, but I wouldn't be surprised if users were confused by it.

Comment (jorisvandenbossche) [reply doc/source/whatsnew/v1.3.0.rst:220]: > I do wonder if this is something in the docs we could find a better solution to - knowing whether to look under GroupBy or DataFrameGroupBy etc

Yeah, I always need to look at https://pandas.pydata.org/docs/dev/reference/groupby.html which one is available to link to ..

In general I think we should try to have most of them available just using `GroupBy` to make life easier. 
We actually do have some duplication as well. Eg for some methods we have both the GroupBy and DataFrameGroupBy version (eg https://pandas.pydata.org/docs/dev/reference/api/pandas.core.groupby.DataFrameGroupBy.all.html#pandas.core.groupby.DataFrameGroupBy.all and https://pandas.pydata.org/docs/dev/reference/api/pandas.core.groupby.GroupBy.all.html#pandas.core.groupby.GroupBy.all). I think for those we can only have the GroupBy version.
